docs/design.md

```diff
@@ -38,13 +58,164 @@ spec:
+
+Process:
+
+Calculate a hash.
+
+Hashes are used to [prevent bucket hot-spots](https://cloud.google.com/blog/products/gcp/optimizing-your-cloud-storage-performance-google-cloud-performance-atlas).
+
+As opposed to a random id, a hash is used to provide a deterministic pattern storing and finding artifacts. By including a cluster name, Substratus can optionally store artifacts across multiple clusters in a single bucket if a compelling use-case arises in the future.
+
+```sh
+hash_input = "clusters/{cluster}/namespaces/{namespace}/{resource_plural}/{name}"
+
+hash = sha256(hash_input)
+```
+
+For example, with the following Model, with a default Substratus installation (cluster = `substratus-1`):
+
+```yaml
+kind: Model
+metadata:
+  name: falcon-7b
+  namespace: team-a
+```
+
+The hash input string would be: `clusters/substratus-1/namespaces/team-a/models/falcon-7b`.
+
+The resulting MD5 hash would be: `f94a0d128bcbd9c1b824e9e5572baf86`.
+
+The following scheme can be used for storing artifacts for Models, Datasets, and Notebooks:
+
+```sh
+# Models
+gs://{bucket}/{hash}/model # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs  # Logs and converted notebooks
+
+# Datasets
+gs://{bucket}/{hash}/data # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs # Logs and converted notebooks
+
+# Notebooks
+gs://{bucket}/{hash}/build/{md5-checksum}.tar # Uploaded build context
+```
+
+The example Model's backing storage would end up being:
+
+```sh
+gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86/model/
+gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86/logs/
+```
+
+The Model would report this URL in its status:
+
+```yaml
+kind: Model
+# ...
+status:
+  url: gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86
+```
+
+#### Reconcile Logic
+
+Pseudo-reconciler logic for a Model:
+
+```
+if .status.url != "" {
+  return
+}
+
+url := "{bucket}/{hash}"
+if sci.bucketObjectExists(url) {
+  # Use the bucket as the source of truth if .status.url did not exist.
+  updateStatus(url)
+  return
+}
+
+# ... Looks like the model should be imported/trained/etc.
+
+runJob()
+```
+
+#### Mount Points
+
+All mount points will are made under a standardized `/home/ml` directory which should correspond to the `WORKDIR` of a Dockerfile. This works well for Jupyter notebooks which can be run with `/home/ml` set as the serving directory: all relevant mounts will be populated on the file explorer sidebar.
```

Comment:
what if my notebook requires root access?

Reply:
You get it now, hurray!